decoder/expression_candidates.go

```diff
@@ -673,6 +673,59 @@ func snippetForExprContraints(placeholder uint, ec schema.ExprConstraints) strin
 	return ""
 }
 
+func snippetForExprContraint(placeholder uint, ec schema.ExprConstraints) string {
+	e := ExprConstraints(ec)
+
+	// TODO: implement rest of these
+	if t, ok := e.LiteralType(); ok {
+		return snippetForLiteralType(placeholder, t)
+	}
+	// case schema.LiteralValue:
+	// 	if len(ec) == 1 {
+	// 		return snippetForLiteralValue(placeholder, et.Val)
+	// 	}
+	// 	return ""
+	if t, ok := e.TupleConsExpr(); ok {
+		ec := ExprConstraints(t.AnyElem)
+		if ec.HasKeywordsOnly() {
+			return "[ ${0} ]"
+		}
+		return "[\n  ${0}\n]"
+	}
+	if t, ok := e.ListExpr(); ok {
+		ec := ExprConstraints(t.Elem)
+		if ec.HasKeywordsOnly() {
+			return "[ ${0} ]"
+		}
+		return "[\n  ${0}\n]"
+	}
+	if t, ok := e.SetExpr(); ok {
+		ec := ExprConstraints(t.Elem)
+		if ec.HasKeywordsOnly() {
+			return "[ ${0} ]"
+		}
+		return "[\n  ${0}\n]"
+	}
+	if t, ok := e.TupleExpr(); ok {
+		// TODO: multiple constraints?
+		ec := ExprConstraints(t.Elems[0])
+		if ec.HasKeywordsOnly() {
+			return "[ ${0} ]"
+		}
+		return "[\n  ${0}\n]"
```

Comment:
Nitpick - also I'm aware that I may be still commenting on WIP here - so you are probably already aware (in which case ignore me! 😄  ) - but the `${0}` can be replaced with a type-aware snippet, as below.

Reply:
I don't understand the reference. I stole/copied this approach from the `snippetForExprContraints` method

Reply:
What I meant is that you can call `snippetForExprContraints` for the inner element of each complex type (SetExpr, ListExpr, first element of TupleExpr & TupleConsExpr). This will lead to e.g.
```hcl
required_attribute = [ "|" ]
```
as opposed to
```hcl
required_attribute = [ | ]
```
(assuming the inner element is of string type)

It's minor thing though - so feel free to ignore! 😉

Reply:
FWIW there's some planned work to generally build support for arbitrarily nested types, so we can address it as part of that task: https://github.com/hashicorp/terraform-ls/issues/496